### PR TITLE
Image and PodSandbox service

### LINKED ISSUE #2: Image runtime interface
Implement the image runtime interface.

### PR DESCRIPTION
**Pull request by nhlfr**
Image and PodSandbox service for virtlet.

Fixes #2 

---
This change is [Reviewable](https://reviewable.io/reviews/mirantis/virtlet/1)

### CODE REVIEW

**Comment by asalkeld** [comment]
some comments in the c code.

---

Review status: 0 of 176 files reviewed at latest revision, 6 unresolved discussions.

---

*[pkg/libvirttools/image.c, line 31 \[r1\]](https://reviewable.io:443/reviews/mirantis/virtlet/1#-KQSmra6fi06GGHN4Ia6:-KQSmra6fi06GGHN4Ia7:bwzv42f) ([raw file](https://github.com/mirantis/virtlet/blob/e2e02bb654b452ad0d1bec08b9490ae94f13d761/pkg/libvirttools/image.c#L31)):*
> ```C
>                        void *opaque)
> {
> 	int *fd = opaque;
> ```

you should really check for NULL opaque before dereferencing it, below.

---

*[pkg/libvirttools/image.c, line 43 \[r1\]](https://reviewable.io:443/reviews/mirantis/virtlet/1#-KQSnF0NdH3e8o5x3bMK:-KQSnF0OdR3Vffn5mxLE:b-14s9e9) ([raw file](https://github.com/mirantis/virtlet/blob/e2e02bb654b452ad0d1bec08b9490ae94f13d761/pkg/libvirttools/image.c#L43)):*
> ```C
> 	int fd = -1;
> 	virStreamPtr stream = NULL;
> ```

If something goes wrong, it's going to be hard to debug. No logs, all errors are the same (-1).

---

*[pkg/libvirttools/image.c, line 69 \[r1\]](https://reviewable.io:443/reviews/mirantis/virtlet/1#-KQSnv-mqs9yKyjaYzo6:-KQSnv-mqs9yKyjaYzo7:b-nl46e0) ([raw file](https://github.com/mirantis/virtlet/blob/e2e02bb654b452ad0d1bec08b9490ae94f13d761/pkg/libvirttools/image.c#L69)):*
> ```C
> 	}
> 
> 	if (close(fd) < 0) {
> ```

i think remove this close() and make use of the one in the cleanup section.

---

*[pkg/libvirttools/image.c, line 74 \[r1\]](https://reviewable.io:443/reviews/mirantis/virtlet/1#-KQSo5ac-OsjVCkMwBTg:-KQSo5ac-OsjVCkMwBTh:byqq93q) ([raw file](https://github.com/mirantis/virtlet/blob/e2e02bb654b452ad0d1bec08b9490ae94f13d761/pkg/libvirttools/image.c#L74)):*
> ```C
> 	}
> 
> 	if (virStreamFinish(stream) < 0) {
> ```

not sure if this is worth even checking, as it goes into cleanup anyways, and do you really want to fail everything if streamfinish fails? the streamfree should clean things up shouldn't it?

---

*[pkg/libvirttools/image.c, line 80 \[r1\]](https://reviewable.io:443/reviews/mirantis/virtlet/1#-KQSmAMEDHei8Eo8Q7L2:-KQSmAMEDHei8Eo8Q7L3:b7s1k1n) ([raw file](https://github.com/mirantis/virtlet/blob/e2e02bb654b452ad0d1bec08b9490ae94f13d761/pkg/libvirttools/image.c#L80)):*
> ```C
> 
>  cleanup:
> 	if (vol)
> ```

braceless if statements are really evil... Just don't go there.

---

*[pkg/libvirttools/image.c, line 83 \[r1\]](https://reviewable.io:443/reviews/mirantis/virtlet/1#-KQSnolROb5laJo5zaaU:-KQSnolROb5laJo5zaaV:b3m52qr) ([raw file](https://github.com/mirantis/virtlet/blob/e2e02bb654b452ad0d1bec08b9490ae94f13d761/pkg/libvirttools/image.c#L83)):*
> ```C
> 		virStorageVolFree(vol);
> 	if (fd)
> 		close(fd);
> ```

isn't this a second close(fd)?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/1#-:-KQSoTcRjzv1W7PrY6bx:b-xf7hvl)*

**Comment by nhlfr** [comment]
Review status: 0 of 176 files reviewed at latest revision, 6 unresolved discussions.

---

*[pkg/libvirttools/image.c, line 43 \[r1\]](https://reviewable.io:443/reviews/mirantis/virtlet/1#-KQSnF0NdH3e8o5x3bMK:-KQU9YYc3-NqtQoNGjUQ:bdqa8ms) ([raw file](https://github.com/mirantis/virtlet/blob/e2e02bb654b452ad0d1bec08b9490ae94f13d761/pkg/libvirttools/image.c#L43)):*
<details><summary><i>Previously, asalkeld (Angus Salkeld) wrote…</i></summary>
> If something goes wrong, it's going to be hard to debug. No logs, all errors are the same (-1).
</details>
1. libvirt prints all its errors to stderr immediately.
2. libvirt stores all its errors and I'm fetching them in go code here:
https://github.com/nhlfr/virtlet/blob/base/pkg/libvirttools/image.go#L207
https://github.com/nhlfr/virtlet/blob/base/pkg/libvirttools/error.go#L31-L39

---

*[pkg/libvirttools/image.c, line 80 \[r1\]](https://reviewable.io:443/reviews/mirantis/virtlet/1#-KQSmAMEDHei8Eo8Q7L2:-KQUAGHUHoeE0C6trnxf:b-elfju6) ([raw file](https://github.com/mirantis/virtlet/blob/e2e02bb654b452ad0d1bec08b9490ae94f13d761/pkg/libvirttools/image.c#L80)):*
<details><summary><i>Previously, asalkeld (Angus Salkeld) wrote…</i></summary>
> braceless if statements are really evil... Just don't go there.
</details>
Why? BTW, libvirt is using braceles if statements too. Eudyptula challenge is teaching me to not use braces in one-liners too. ;)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/1)*

**Comment by nhlfr** [comment]
Review status: 0 of 176 files reviewed at latest revision, 6 unresolved discussions.

---

*[pkg/libvirttools/image.c, line 31 \[r1\]](https://reviewable.io:443/reviews/mirantis/virtlet/1#-KQSmra6fi06GGHN4Ia6:-KQUFQMN14icTZg4IAtL:b-896fix) ([raw file](https://github.com/mirantis/virtlet/blob/e2e02bb654b452ad0d1bec08b9490ae94f13d761/pkg/libvirttools/image.c#L31)):*
<details><summary><i>Previously, asalkeld (Angus Salkeld) wrote…</i></summary>
> you should really check for NULL opaque before dereferencing it, below.
</details>
Done.

---

*[pkg/libvirttools/image.c, line 69 \[r1\]](https://reviewable.io:443/reviews/mirantis/virtlet/1#-KQSnv-mqs9yKyjaYzo6:-KQUFUw-YkJskfcHAbV8:b-896fix) ([raw file](https://github.com/mirantis/virtlet/blob/e2e02bb654b452ad0d1bec08b9490ae94f13d761/pkg/libvirttools/image.c#L69)):*
<details><summary><i>Previously, asalkeld (Angus Salkeld) wrote…</i></summary>
> i think remove this close() and make use of the one in the cleanup section.
</details>
Done.

---

*[pkg/libvirttools/image.c, line 74 \[r1\]](https://reviewable.io:443/reviews/mirantis/virtlet/1#-KQSo5ac-OsjVCkMwBTg:-KQUF1fZ7VvmbB9txek5:bx7qt0b) ([raw file](https://github.com/mirantis/virtlet/blob/e2e02bb654b452ad0d1bec08b9490ae94f13d761/pkg/libvirttools/image.c#L74)):*
<details><summary><i>Previously, asalkeld (Angus Salkeld) wrote…</i></summary>
> not sure if this is worth even checking, as it goes into cleanup anyways, and do you really want to fail everything if streamfinish fails? the streamfree should clean things up shouldn't it?
</details>
virStreamFinish fails when data weren't successfully written - and it saves this error in libvirt internals which I mentioned above, so we can receive this exact error later. And virStreamFinish doesn't do cleaning at all.

---

*[pkg/libvirttools/image.c, line 83 \[r1\]](https://reviewable.io:443/reviews/mirantis/virtlet/1#-KQSnolROb5laJo5zaaU:-KQUFW2rVKm_oZ37EKF5:b-896fix) ([raw file](https://github.com/mirantis/virtlet/blob/e2e02bb654b452ad0d1bec08b9490ae94f13d761/pkg/libvirttools/image.c#L83)):*
<details><summary><i>Previously, asalkeld (Angus Salkeld) wrote…</i></summary>
> isn't this a second close(fd)?
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/1)*

File: contrib/docker/libvirt/start.sh [hunk 1]
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+
+set -o errexit
+
+modprobe kvm
```
Comment (jellonek) [inline contrib/docker/libvirt/start.sh]: This seems to be problematic inside of container. Did you tested this with running `modprobe -r kvm` on host before starting this dock?

Comment (jellonek) [inline contrib/docker/libvirt/start.sh]: Ok, this also could be used as check, but still intention seems to be not specified - maybe describe this in way similar to:
```
modprobe kvm || echo "Missing kvm module on host" && exit 1
```

Comment (nhlfr) [inline contrib/docker/libvirt/start.sh]:
```
$ sudo modprobe -r kvm
$ lsmod | grep kvm    
$ docker-compose up -d
Starting dockercompose_libvirt_1
Starting dockercompose_virtlet_1
$ lsmod | grep kvm    
kvm                   561152  0
irqbypass              16384  1 kvm
```

Comment (nhlfr) [inline contrib/docker/libvirt/start.sh]: However, I'll try to handle kvm_intel and kvm_amd modules as well.

File: pkg/libvirttools/image.c [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+		result = -1;
+		goto cleanup;
+	}
+
+	if (virStreamSendAll(stream, virtletVolUploadSource, &fd) < 0) {
+		result = -1;
+		goto cleanup;
+	}
+
+	if (virStreamFinish(stream) < 0) {
```
Comment (jellonek) [inline pkg/libvirttools/image.c]: Please combine this multiple ifs into single one with `||` conditions (multilined) what will remove need of `goto` usage.

Comment (nhlfr) [inline pkg/libvirttools/image.c]: I did that where possible, but I couldn't get rid of `goto` at all - meanwhile I realized that @asalkeld was partially right about returning different error statuses. There is one place (opening the file descriptor) where I decided to return `errno`.

**Comment by jellonek** [comment]
:lgtm:

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/1#-:-KQVQfNJx7toQ5owe4dW:bnfp4nl)*

**Comment by asalkeld** [comment]
The C code and calling GO code LGTM.

**Comment by asalkeld** [comment]
Reviewed 173 of 176 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/1)*

**Comment by asalkeld** [comment]
:lgtm:

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/1#-:-KQZK2jKIdS_zEFfORAM:bnfp4nl)*

**Pull request merged**